bytecomp/bytesections.ml

```diff
@@ -15,31 +15,94 @@
 
 (* Handling of sections in bytecode executable files *)
 
-(* List of all sections, in reverse order *)
-
-let section_table = ref ([] : (string * int) list)
+module Name = struct
+
+  type raw_name = string
+
+  type t =
+    | CODE (** bytecode *)
+    | CRCS (** crcs for modules *)
+    | DATA (** global data (constant) *)
+    | DBUG (** debug info *)
+    | DLLS (** dll names *)
+    | DLPT (** dll paths *)
+    | PRIM (** primitives names *)
+    | RNTM (** The path to the bytecode interpreter (use_runtime mode) *)
+    | SYMB (** global identifiers *)
+    | Other of raw_name
+
+  let of_string name =
+    match name with
+    | "CODE" -> CODE
+    | "DLPT" -> DLPT
+    | "DLLS" -> DLLS
+    | "DATA" -> DATA
+    | "PRIM" -> PRIM
+    | "SYMB" -> SYMB
+    | "DBUG" -> DBUG
+    | "CRCS" -> CRCS
+    | "RNTM" -> RNTM
+    | name   ->
+        if String.length name <> 4 then
+          invalid_arg "Bytesections.Name.of_string: must be of size 4";
+        Other name
+
+  let to_string = function
+    | CODE -> "CODE"
+    | DLPT -> "DLPT"
+    | DLLS -> "DLLS"
+    | DATA -> "DATA"
+    | PRIM -> "PRIM"
+    | SYMB -> "SYMB"
+    | DBUG -> "DBUG"
+    | CRCS -> "CRCS"
+    | RNTM -> "RNTM"
+    | Other n -> n
+end
+
+type section_entry = {
+  name : Name.t;
+  pos  : int;
+  len  : int;
+}
+
+type section_table = {
+   sections : section_entry list;
+   first_pos : int
+}
 
 (* Recording sections *)
-
-let section_beginning = ref 0
-
-let init_record outchan =
-  section_beginning := pos_out outchan;
-  section_table := []
-
-let record outchan name =
+type toc_writer = {
+  (* List of all sections, in reverse order *)
+  mutable section_table_rev : section_entry list;
+  mutable section_prev : int;
+  outchan : out_channel;
+}
+
+let init_record outchan : toc_writer =
   let pos = pos_out outchan in
-  section_table := (name, pos - !section_beginning) :: !section_table;
-  section_beginning := pos
-
-let write_toc_and_trailer outchan =
+  { section_prev = pos;
+    section_table_rev = [];
+    outchan }
+
+let record t name =
+  let pos = pos_out t.outchan in
+  if pos < t.section_prev then
+    invalid_arg "Bytesections.record: out_channel offset moved backward";
+  let entry = {name; pos = t.section_prev; len = pos - t.section_prev} in
+  t.section_table_rev <- entry :: t.section_table_rev;
+  t.section_prev <- pos
+
+let write_toc_and_trailer t =
+  let section_table = List.rev t.section_table_rev in
   List.iter
-    (fun (name, len) ->
-      output_string outchan name; output_binary_int outchan len)
-    (List.rev !section_table);
-  output_binary_int outchan (List.length !section_table);
-  output_string outchan Config.exec_magic_number;
-  section_table := [];
+    (fun {name; pos = _; len} ->
+       let name = Name.to_string name in
+       assert (String.length name = 4);
```

Comment:
```suggestion
```

Reply:
I would be a bit careful here. I think that the `assert` is not meant to prevent a failure condition or explain a size manipulation in this function, but rather to justify the constant `4` that appears in the matching `read_toc` function below. You can read both in parallel and check that one inverts the other. Now we could avoid the `4` appearing on the other end (notably by using `Name.read ic` instead of `Name.of_string (really_input_string ic 4)`, but as long as one is there we probably want to keep the other.

Reply:
Good point, let's disregard this suggestion.